Bayesian SINDy

mikkelbue opened the pull request:
Hi all!

This is a (possible) implementation of Bayesian SINDy. It uses `numpyro` under the hood to drive NUTS sampling.

- It is implemented as a `pysindy.optimizers` module, `SBR` for Sparse Bayesian Regression.
- Uses the regularised horseshoe prior outlined in [Hirsh, et al.](https://arxiv.org/abs/2107.02107).
- Parameters to the `SBR` optimizer include hyperparameters to the statistical model, i.e. `tau_0`, `nu` and `s` for the hyperprior and `lamb` for the exponential prior over the measurement noise.
- Additional kwargs are fed forward to the `numpyro.infer.MCMC` object.
- `self.coef_` is set to the **mean** of the posterior samples.
- Additional MCMC diagnostics can be invoked by inspecting the `SINDy.optimizer.mcmc` object using `arviz`.
- New notebook [16_bayesian_sindy.ipynb](https://github.com/mikkelbue/pysindy/blob/bayesian_sindy/examples/16_bayesian_sindy.ipynb) demonstrates the method.

This PR is a work in progress in that test and docs are missing. But I thought it would be better to put it in early in the process, in case anyone has questions, ideas or comments with regards to the implementation. Please let me know if this is a bad idea.

Thanks for considering this PR.

TODO:
- [x] Documentation
- [ ] Tests
- [x] Improve / polish notebook

Jacob-Stevens-Haas commented:
Thank you so much for this - I've been interested in getting Hirsch et al's spike slab priors into pysindy for some time!  And selfishly, I've wanted a reason to start interacting with jax.  We implemented a bit of kernel smoothing from a paper of @jfelipeosorio in [derivative](https://github.com/andgoldschmidt/derivative/pull/30), but we shied away from anything that required jax.

Giving this a quick look now (but not yet a full review), here's some more engineering, less mathematical considerations:
**Import**
* Add an `[project.optional-dependencies]` section for the dependencies
* You noticed the import guard that we did in SR3 for the optional cvxpy dependencies, including the PEP 690 comment.  I just checked and it appears PEP 690 was rejected, so you can remove the comment.  Someone (maybe me?) needs to do some reading on what the right way to guard import statements.

**Example**
* There already is an example 16 at this point
* See the [guidance](https://github.com/dynamicslab/pysindy#contributing-examples) on creating jupyter notebook examples. TL;DR: Maintaining compatibility with existing notebooks is tough.  Its not uncommon for me to run the whole test suite between ten and a hundred times per PR, which means I'm not running it on notebooks.  Obviously unit tests make this easier, but if you want to ensure the notebook itself stays runnable, we need to test it with [this code](https://github.com/dynamicslab/pysindy/blob/66112bb544f8907c3ae91e30d3ba2a42d5a1413b/test/test_notebooks.py#L67) in order for CI to care about it.  You'd also need to generate the ipynb with examples/publish_notebook.py. (This part is pretty easy - save the notebook as a .py file, then run `python publish_notebook.py XX_Bayesian_SINDy`)
* I like the example a lot - it shows "how to use this" rather than "here's all the different ways you can use this on a variety of systems".  We have some people in the lab working on a "Bayesian" SINDy that would definitely appreciate this.
* Include some LaTeX for the equations
* I assume the rows of the final plot are terms in the discovered equation - could you label them?

**Code**
* Can you add type annotations and docstrings?
* What is the difference between `SBR` and `SparseBayesianRegression`?  I can tell the latter is used in the former, but the former being just the abbreviation of the latter doesn't give any indication of the role that these objects play.  Maybe rename `SBR` -> `SBROptimizer`, but you're probably in the best place to suggest names.
* Best practice is not necessarily settled, but my stance on code variables that shadow math variables: `meaning_letter`.  e.g. if someone's doing L-1 regression, instead of naming a variable `lambda`, it would be `reg_coef_lambda`.  It helps people who have read the paper but not recently (me in this example) to still be able to troubleshoot errors.

**Ok I lied, some math**
* It's been a while since I read Hirsch, but is this mathematically compatible with `EnsembleOptimizer`, which subsamples the data and/or library, wraps any `BaseOptimizer`, and takes the mean/median of resulting models?  My feeling is no, which indicates that out typing system needs to expand so as to prohibit this case.

mikkelbue commented:
Hi @Jacob-Stevens-Haas 

Thanks for your comments and suggestions. I'm happy if this is useful! 

I have made a few changes:

- Added `numpyro` as an optional dependency.
- Removed reference to PEP 690.
- Renamed `tau_0 -> sparsity_coef_tau0`, `nu -> slab_shape_nu`, `s -> slab_shape_s`, `lamb -> noise_hyper_lambda`.
- Renumbered the notebook 16 -> 19
- Moved the methods from `SparseBayesianRegression` into `SBR`. I set it up like that because I thought it was neat to have a general-purpose sparse Bayesian regressor, but it's simpler to just make everything part of the optimizer. I'm not precious about naming, just thought `SBR` would fit into the existing convention.

I'll have a look at your other points when I have more time.

mikkelbue commented:
Also, just as a point of order:

This implementation uses the _regularised horseshoe_ prior, not (yet) the _spike and slab_ prior. The regularised horseshoe is computationally more convenient because it is differentiable, so you can just plug the whole shebang into a NUTS sampler.

The spike and slab prior AFAIK requires a compound step sampler or some other clever way of making categorical moves. It's something I can look into, but I am not currently sure what is the best way to achieve this in `numpyro`.

Jacob-Stevens-Haas commented:
Oh, sorry, yes I remember now.  I meant to say horseshoe prior.  I didn't mean to point you in a different direction.

mikkelbue commented:
@Jacob-Stevens-Haas quick update. 2 new commits:
1. Optimized the sampling of coefficients by vectorizing (seems obvious now, haha)
2. Added docstring with type hints.

Still looking to update the notebook with the stuff you requested.

mikkelbue commented:
@Jacob-Stevens-Haas Okay, I think this addresses all of your points. Please let me know if anything can be improved.

I'm not sure what to add in terms of tests. The notebook has a `testing` mode now so that should cover some of it.

Jacob-Stevens-Haas commented:
Thank you!  I'll go ahead and review it later this week.  Before then, I'll figure out how to get it so your CI runs don't need manual approval.  Sorry to keep you waiting.

For tests, 
1. look at `test_optimizers` for tests that are parametrized to all optimizers
2. Think about which `ValueErrors` (bad arguments) you want to check for.
3. Is there a simple very-low-dimensional (like a length 2, 3, or 4 vector) regression where it's easy to see how the math for horseshoe priors should work out?  The three tests starting here are [good examples](https://github.com/dynamicslab/pysindy/blob/638e4bb1657ee9a7caaeca753ec0749f32f194fa/test/test_optimizers.py#L1088).
4. Are there any helper functions that you use that have a simple test?  [Here's one](https://github.com/dynamicslab/pysindy/blob/638e4bb1657ee9a7caaeca753ec0749f32f194fa/test/test_optimizers.py#L1124) where we pulled out an indexing subroutine, and therefore could test it independently.

Jacob-Stevens-Haas commented:
Ok I added you as a contributor with "write" permission to the repo.  You may need to click an invite link in an email, but that should enable CI to run automatically on your pushes.

mikkelbue commented:
@Jacob-Stevens-Haas Thank you!

The tests are failing when building the docs and I can't seem to figure out why. This is the trace:
```
sphinx.errors.SphinxWarning: autodoc: failed to import module 'sbr' from module 'pysindy.optimizers'; the following exception was raised:
No module named 'jax'

Warning, treated as error:
autodoc: failed to import module 'sbr' from module 'pysindy.optimizers'; the following exception was raised:
No module named 'jax'
```

I have already included `jax` as an optional dependency in the `numpyro` group, so I don't understand how this is different from the way `cvxpy` is set up for `SINDyPI`. But I must have missed something.

Jacob-Stevens-Haas commented:
The github workflow file creates a new environment and installs the current package, including **all specified optional dependencies**.  You'd need to change [this line](https://github.com/dynamicslab/pysindy/blob/638e4bb1657ee9a7caaeca753ec0749f32f194fa/.github/workflows/main.yml#L43C20-L43C20) to add your optional dependencies to the list.

Jacob-Stevens-Haas commented:
Hey sorry for slow responses on my end - I'm responsible for #451 and #459, which are both pretty complex.  I'll finish #451 in a few days and then I'll give this a thorough review.

Jacob-Stevens-Haas reviewed (changes requested):
Looking good!  Mostly docstring stuff, otherwise, quality of life edits.

I did a bit of reading and math to grok `s` and `nu`, leaving docstring suggestions.  Please do assess whether my wording suggestions make sense with how you understand the horseshoe prior.  As for tests, I'll follow up with a list of existing tests to add `SBR` to.  For the notebook, save it as a .py file, add the .py file to the repo, and I can show you in review how to edit it to make it fast testable.

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+        slab_shape_nu=4,
+        slab_shape_s=2,
+        noise_hyper_lambda=1,
+        normalize_columns=False,
+        copy_X=True,
+        **mcmc_kwargs,
```
Okay. I wouldn't use [mutable default arguments](https://docs.python-guide.org/writing/gotchas/), but that is of course up to you. A safer way to do this would be like this:
```
def __init__(self, mcmc_kwargs=None):
    if mcmc_kwargs is None:
        mcmc_kwargs = {}
```

Jacob-Stevens-Haas replied:
Oh god point about the mutable default arguments - yeah you have the right way of doing it.

mikkelbue commented:
Thanks for all your comments. I responded to one just now, while I had the thought. I'll address the remaining sometime next week.

mikkelbue commented on pyproject.toml:
```diff
@@ -61,6 +61,10 @@ cvxpy = [
     "cvxpy",
     "scs>=2.1, !=2.1.4"
 ]
+numpyro = [
+    "numpyro",
+    "jax"
+]
```
OK.

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+    coef_ : array, shape (n_features,) or (n_targets, n_features)
+        Posterior means of the SINDy coefficients.
+
+    mcmc : numpyro.infer.MCMC
```
@Jacob-Stevens-Haas do you mean that `mcmc` should also have a trailing underscore, like `mcmc_`?

Jacob-Stevens-Haas replied:
Yep, exactly

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+
+    Parameters
+    ----------
+    sparsity_coef_tau0 : float, optional (default 0.1)
+        Sparsity coefficient for regularised horseshoe hyper-prior. Lower
+        value increases the sparsity of the SINDy coefficients. See Hirsh
+        et al. (2021) for more details.
+
+    slab_shape_nu : float, optional (default 4)
+        Shape parameter for the slab region of the regularised horseshoe
+        prior. See Hirsh et al. (2021) for more details
+
+    slab_shape_s : float, optional (default 2)
+        Shape parameter for the slab region of the regularised horseshoe
+        prior. See Hirsh et al. (2021) for more details
+
+    noise_hyper_lambda : float, optional (default 1)
+        Rate hyperparameter for the exponential prior distribution over
+        the noise standard deviation.
```
@Jacob-Stevens-Haas I'm happy with this. The `noise_hyper_lambda` is the hyperparameter for the exponential distribution over the observation noise parameter `sigma`. So it's not relevant to the half-cauchys for the parameters, only the Gaussian likelihood for the data.

Jacob-Stevens-Haas replied:
Thanks!  OK I see equation 3.3/3.4 "We must choose priors for the noise level parameters $σ_j$".  

Soooo this gets dramatically more exciting, but the tl;dr: is that the docstring for it is great, maybe just name `noise_hyper_lambda` something like `noise_hyper`, since the paper didn't use a greek letter here.  I'll have some more questions about the MCMC implementation, but maybe dial the default way down (is zero legal?), since (a) the case of recovering an exact simulation without noise is perhaps the most common, and (b) if i read it correctly (check me here) the paper uses different hypers & distributions for measurement variance prior: $\sigma_u, \sigma_\nu \sim LogNormal(\mu=-1, \sigma=.1)$ in § 3.2.1, $\sigma_u, \sigma_\nu \sim Gamma(\alpha=1, \beta=1)$ in § 3.2.2, and unspecified in § 3.2.3.  What are your thoughts?

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+
+        # run the MCMC
+        kernel = NUTS(self._numpyro_model)
+        num_warmup = kwargs.pop("num_warmup", 2000)
+        num_samples = kwargs.pop("num_samples", 5000)
```
I'll have a look at this.

mikkelbue replied:
This is fixed in https://github.com/mikkelbue/pysindy/commit/cacb02b7be3601a95f3d32fb0732146874776335

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+        copy_X=True,
+        **mcmc_kwargs,
+    ):
+        super().__init__(
+            copy_X=copy_X,
+            normalize_columns=normalize_columns,
+        )
```
I'll have a look at this.

mikkelbue replied:
This is fixed in https://github.com/mikkelbue/pysindy/commit/cacb02b7be3601a95f3d32fb0732146874776335

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+    def _run_mcmc(self, x, y, **kwargs):
+        # set up a jax random key.
+        seed = kwargs.pop("seed", 0)
+        rng_key = random.PRNGKey(seed)
+        rng_key, rng_key_ = random.split(rng_key)
```
It's a JAX-specific thing, please see here: [JAX - The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers)

I'll have a look and see if I can make it more readable.

mikkelbue replied:
I guess it is not strictly necessary, since there are no other seeded operations after `mcmc.run`.

Jacob-Stevens-Haas replied:
Thanks, that's a great link, I think i get it.  Since we're only consuming one key, and `random.split(0)` is just as random as `0`, could we just use the `rng_key = random.PRNGKey(seed)`?  line?

mikkelbue replied:
Absolutely. I think this was a leftover from some early experiments. This is addressed in https://github.com/dynamicslab/pysindy/pull/440/commits/952b98932c8e4faaf5fbe7313768bd276f34a4c7

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+        return mcmc
+
+
+def sample_reg_horseshoe_hyper(tau0=0.1, nu=4, s=2):
+    """
+    For details on this prior, please refer to:
+    Hirsh, S. M., Barajas-Solano, D. A., & Kutz, J. N. (2021).
+    parsifying Priors for Bayesian Uncertainty Quantification in
+    Model Discovery (arXiv:2107.02107). arXiv. http://arxiv.org/abs/2107.02107
+    """
+    tau = numpyro.sample("tau", HalfCauchy(tau0))
+    c_sq = numpyro.sample("c_sq", InverseGamma(nu / 2, nu / 2 * s**2))
+    return tau, c_sq
```
OK

mikkelbue replied:
This is addressed in https://github.com/mikkelbue/pysindy/commit/7af3848b87790cf2880a495507fd69e10dd28cc9

mikkelbue commented on .github/workflows/main.yml:
```diff
@@ -40,7 +40,7 @@ jobs:
       run: |
         pip install --upgrade pip
-        pip install .[dev,miosr,cvxpy]
+        pip install .[dev,miosr,cvxpy,numpyro]
```
Fixed in https://github.com/mikkelbue/pysindy/commit/f0a682338cef351dd2c71c9a5d253597c40b5e76

mikkelbue commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,163 @@
+
+    def __init__(
+        self,
+        sparsity_coef_tau0=0.1,
+        slab_shape_nu=4,
+        slab_shape_s=2,
+        noise_hyper_lambda=1,
```
@Jacob-Stevens-Haas okay, can I do that. I just thought that since none of the other optimizers have type hints in the `__init__` method, `SBR` shouldn't have that either.

Jacob-Stevens-Haas replied:
Totally makes sense... we'd like to have them throughout, and I've been adding them bit by bit in PRs, which is why there's inconsistent practice throughout.  But it's a huge help in the IDE when refactoring when they exist.

More of a nice-to-have at this point, though.

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -76,22 +76,27 @@ def __init__(
+        else:
+            self.mcmc_kwargs = {}
 
     def _reduce(self, x, y):
```
Is `y` is expected to be measurements of the system, in [this line](https://github.com/dynamicslab/pysindy/blob/01698bee0a6ed69209d132fca065b6f0ec68a89d/pysindy/optimizers/sbr.py#L128)?  In general SBR, as a SINDy optimizer, the `y` argument to `_reduce` will receive the estimated `x_dot` from whatever differentiation method is used.  Callers can skip the differentiation step by directly passing derivatives, e.g. `SINDy(optimizer=SBR()).fit(x=x, x_dot=x)`.  If I'm understanding this correctly, we should mention that in the docstring.

Long term, I have a `sss` branch in which I hope to implement a more clear `SINDy` subclass for these kinds of problems that eschew a differentiation method.

mikkelbue replied:
@Jacob-Stevens-Haas that's right, `y` is expecting derivatives `x_dot`, like the other optimizers.

The line `numpyro.sample("obs", Normal(mu, sigma), obs=y)` is the likelihood, and it says that the errors of these derivatives are Gaussian. It's maybe a little misleading that the parameter of the likelihood is called `obs`, but that's kind of a quirk the Bayesian formalism.

mikkelbue replied:
@Jacob-Stevens-Haas I think I might have made a massive blunder and overlooked something in Hirsh's paper.

The observation model and likelihood is with respect to the states in Hirsh, not the derivatives. The way I have implemented it, the likelihood is with respect to the derivatives.

I don't think it will matter much in terms of the actual numerical results for the coefficients, especially since the noise parameter `sigma` is free. However, the current code does not do what is described in Hirsh's paper.

I think what we would actually want is to be able to reconstruct the states from within the `_numpyro_model`, basically by running something equivalent of `pysindy.simulate`, i.e.

I don't know what is the best thing to do here, I'll have to think this through.

```
# sample the parameters compute the predicted outputs.
beta = _sample_reg_horseshoe(tau, c_sq, (n_targets, n_features))

# get SINDy state prediction given coefficients beta
x_hat = model.simulate(beta)

# compute the likelihood.
sigma = numpyro.sample("sigma", Exponential(self.noise_hyper_lambda))
numpyro.sample("obs", Normal(x_hat, sigma), obs=y)
```

Jacob-Stevens-Haas replied:
I wouldn't worry too much.  There's a few other places in pysindy that the choice of one component requires the user to handle things differently outside of that component: `TrappingSR3` optimizer requires the user to choose a degree-2 polynomial library, and `WeakPDELibrary` means that the `SINDy` the differentiation method ends up as a no-op.  In the latter case, we wrote a hook for feature libraries to control how differentiation is applied, but that's little more than a one-time hack that we can't do here.

Short term, we'll just have to expect user to know to only use `SBR` in cases where `model.fit(x=x, x_dot=x)`.  I think the PR has enough merit to users, since users of bleeding-edge SINDy are almost exclusively researching the methods themselves, rather than applying them to real world data.  If it's restrictions are clear in the docstring, I'm happy.

Long term, this is discussed in #351 and #407, which I'll start working on once I've cleaned up the trapping PRs.  This PR is a great template for how they should be implemented.

Jacob-Stevens-Haas replied:
Ohh, I see I misunderstood the role of `mu = jnp.dot(x, beta.T)`.  Yeah, the paper does simulate.  Maybe in place of 3.3/3.4 we use the random variable that equates to standard SINDy, i.e. $\dot x \sim$ Normal $(\Xi^T\Theta(x), \sigma))$, but keep the prior $\sigma\sim$ Exp $(\lambda)$?

mikkelbue replied:
@Jacob-Stevens-Haas Thanks for looking into this.

I have been doing a bit of digging, and it will be significantly more complex to do what is in the paper. Mainly because `jax` tracks operations to compute derivatives. This is not compatible with e.g. `pysindy.feature_library` (which we would need for the `predict` step in the simulation), where e.g. `np.asarray` is employed for type coercion.

Similarly, `scipy.integrate.solve_ivp` also employs type coercion, but to use the NUTS sampler in `numpyro`, gradients would have to be tracked by `jax` through the numerical simulation. There is a `jax`-compatible ODE solver called [`diffrax`](https://github.com/patrick-kidger/diffrax) which could be employed to this end, but that is a relatively complex software engineering task.

I think the current implementation is valid for `discrete_time` simulations, which we could make clear in the documentation. I would be interested in looking into a continuous-time implementation too, using e.g. `diffrax` for simulating the forward model, but I think that's a bit of a longer term goal.

Jacob-Stevens-Haas replied:
Ok, thanks.  Yeah, long term diffrax would be great to have, as would removing all calls to `np.asarray`.  Relatedly, this repo also maintains a subclass of numpy array (See #451) to allow axis labels to propagate through algorithms, which is super useful in debugging, but it's stripped whenever we use `np.asarray`

mikkelbue replied:
> Relatedly, this repo also maintains a subclass of numpy array (See https://github.com/dynamicslab/pysindy/pull/451) to allow axis labels to propagate through algorithms,

Yes, I have noticed that. Do you know of [Xarray](https://xarray.dev/)? I think they are doing something similar.

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -76,22 +76,27 @@ def __init__(
-            copy_X=copy_X,
-            normalize_columns=normalize_columns,
-        )
+        super().__init__(**kwargs)
```
Since optimizers all inherit the `unbais` kwarg and `_unbias()` step from `BaseOptimizer`, and that step appears incompatible with SBR, I think we should explicitly add `unbias: bool=False` to the init args here, guard `if unbias: raise ValueError` with a message like [this one](https://github.com/dynamicslab/pysindy/blob/a7eb8901f1e333273ef8ec17ad6dab61db15c2ab/pysindy/optimizers/trapping_sr3.py#L326), and pass `unbias` explicitly in the superclass constructor`.

mikkelbue replied:
Okay dokey. This is fixed in https://github.com/dynamicslab/pysindy/pull/440/commits/28dd0f3ac1675c31f732d0c0a7791b3305107c28

Jacob-Stevens-Haas reviewed (changes requested):
Thanks for keeping on this, and thanks for your patience.  I really appreciate how it's keeping me close to the literature, and I'm learning a lot.  Some of my comments may be wrong if I've misunderstood the material so definitely push back if I'm confusing things.


It appears the tests fail because numpyro 0.13 drops support for python 8;  I was wrong to suggest ~=13.2.  >=0.12 should probably be fine.

mikkelbue commented:
Hi @Jacob-Stevens-Haas 

I'm still working on this, just a little stuck on the next step. I'd like to include an example where the algorithm is used with `discrete_time`, as this is the case where the Bayesian error model relates directly to the states, [as discussed above](https://github.com/dynamicslab/pysindy/pull/440#discussion_r1463528279).

But I am struggling to find a good example. I haven't worked much with discrete time equations before. It's easy enough to take a differential equation and put it into difference equation form. But then you are essentially just doing some Euler, and the examples I have tried so far tend to diverge very fast, unless I use very short timesteps. But that is maybe fine?

Please have a look at [this example]( https://github.com/mikkelbue/pysindy/blob/bayesian_sindy/examples/19_bayesian_sindy/example-discrete.ipynb) and let me know it makes sense. It's just a harmonic oscillator but rearranged as a difference equation with first order Euler integration.

Thanks!

Jacob-Stevens-Haas commented:
Sorry I didn't respond - I think it's a good example.  The only other common discrete system I know of is the logistic map from example 1, but either is fine.  That said, I just realized that technically `discrete_time=True` might have the same issue as continuous; wouldn't the integral term for $\hat x_j$ in eq 3.3 be replaced with a sum term across timepoints until $t_i$?

Again though, I still think it's valuable in both cases, it just isn't the "full-Bayesian" form in the paper.  The difference is that the random variable in eq 3.3 is now also conditioned on the calculated derivatives from the first step of SINDy.  While in reality, they're those derivatives depend of $X$, we just treat them as independent measurements in the second step.

mikkelbue commented:
Hi @Jacob-Stevens-Haas 

No worries at all. I am aware that I have gone ahead and created extra work for you - without asking first :laughing: 

> That said, I just realized that technically discrete_time=True might have the same issue as continuous; wouldn't the integral term for in eq 3.3 be replaced with a sum term across timepoints until?

If the system is observed at all timesteps, then you shouldn't have to worry about that summation. The terms of the sum are baked into the right-hand side of the difference equation `discrete_time`  formulation. So you march the system forward with $x_{t+1} = f(x_t, t)$, with whichever terms you are adding together already part of $f(x_t, t)$. As far as I understand, anyway.

Apparently, I am a complete masochist, so I have gone ahead and made the Hirch formulation work with `diffrax` as an integrator. This is on a [separate branch](https://github.com/mikkelbue/pysindy/tree/bayesian_sindy_diffrax) if you want to have a look. It's still pretty sketchy but it does work.
Some challenges:
1. As discussed, the `feature_library` classes do not handle `jax.numpy.array`s, so I have implemented an unsafe version of the `PolynomialFeatures` for use with the integrator.
2. It only handles a single trajectory for the time being. It shouldn't be difficult to generalise to multiple trajectories though.
3. I don't think it handles control variables yet, but I haven't tested it.
4. It is far more unstable than the other implementation and does not converge unless given sensible starting guess. Which is not that surprising.
5. It is also a lot slower, since it needs to do that integration for each MCMC sample.

This [example](https://github.com/mikkelbue/pysindy/blob/bayesian_sindy_diffrax/examples/19_bayesian_sindy/example.ipynb) now uses that formulation. If I can get this new branch polished up a bit, I would consider splitting the optimizer into two: `SBR` and `LiteSBR` to allow using both, since they both have are pros and cons.

What are your thoughts? This would take a little more work to get over the finish line, and the solution is a little less clean than the first one, but it might be useful despite its limitations.

mikkelbue commented:
Quick note: I have opted to make the existing "approximate" method available through a flag `exact=False` in the `__init__` method of `SBR`, rather than a separate `LiteSBR` optimizer. All this is still on a [separate branch](https://github.com/mikkelbue/pysindy/blob/bayesian_sindy_diffrax/).

I think this is a pretty robust implementation at this point. The only glaring issues are
1. The adaptive step size controller `PIDController` from `diffrax` does not seem to play well with `numpyro.infer.MCMC`, and a `XlaRuntimeError` is raised every time I attempt to run use it. I can't figure out the reason for this issue.
2. This implementation is so far only compatible with `PolynomialLibrary` and whichever feature library we choose to "port" to the `ProxyFeatureLibrary`.

Jacob-Stevens-Haas commented:
:sweat_smile: Ok this one's a thinker.  Long story short, let's focus on this PR first.  Long story still too short but a bit longer:

One of the thorniest issues I've dealt with in pysindy is the multiplicity of classes that involve repeating code to make a particular use case or method work, often signaled by `if isinstance(...` calls.  It makes it really hard to know how to combine methods in code or which methods are mathematically compatible, and subtle changes to the internal API cause breakages in unexpected locations.  A better approach is to create the type system that works and enforces mathematical sensibility.  It's better to start with discussing what about the feature library and optimizer API is missing to enable this method as an Issue.

E.g. Are "feature library" and "optimizer" even the right abstractions to use, or is another layer implicit?  Off the top of my head, this seems like it could be solved with a mixin class for _all_ feature libraries (except maybe WeakPDELibrary, another example of a class not fitting into the existing abstractions but being forced to).  But a mixin may be the wrong solution.  Maybe it also requires cooperative multiple inheritance for both optimizers and feature libraries, and the `SINDy` class determines which direction feature libraries to inherit (default or diffrax) at runtime?

<hr>

As for the discrete stuff, I was wrong about the sum - it's composition!  It took a while to figure out, and I'm not 100% sure I have it right. But the key for me was figuring out the subtly confusing notation in eq 3.3: Way above it, Hirsh et al. introduce $\mathbf{X = [y_1, ..., y_n}]^T$ as a measurement.  I missed that and had stupidly assumed that a scalar element of $\mathbf{X}$ was $x_{ij}$, but it's actually $y_{ij}$ :sob:   

But anyways, to illustrate my point, here's Eq 3.3, the data likelihood (ignoring $j$ and $\phi$ WLOG):
$$p(y_i | \mathbf{\Xi}, x_0) = \frac{1}{...} \|\|y_i - \hat x_i\|\|^2 = \frac{1}{...}\|\|y_i - x_0 + \int_0^{t_i} \mathbf{\Xi} \Theta(x(t')) dx'\|\|^2$$
$$\prod_i p(y_i | \mathbf{\Xi}, x_0)= p(\mathbf{X|\Xi}, x_0)$$

For modeling a discrete system, it would be:

$$p(y_i | \mathbf{\Xi}, x_0) = \frac{1}{...}\|\|y_i - \hat x_i\|\|^2 = \frac{1}{...}\|\|y_i - \mathbf{\Xi}\Theta(...\mathbf{\Xi}\Theta(x_0)...)\|\|^2$$

with the same product expression holding.   I think we've both seen that the [related line in the SBR code](https://github.com/dynamicslab/pysindy/blob/ab2641b36d89a5358cfb52f2b731457d3193b12f/pysindy/optimizers/sbr.py#L146), paraphrased here:
```python
numpyro.sample("obs", Normal(jnp.dot(x, beta.T), sigma), obs=y)
```
instead represents the standard SINDy data likelihood (with `beta` = $\mathbf{\Xi}$, `x`= $\Theta(X)=[..., \Theta(y_i),...]$, `y`= $\dot X =[..., \dot y_i,...]$) , 
$$p (\dot{y_i}|\mathbf{\Xi},  y_i) = \frac{1}{\dots}\|\|\dot y_i -\mathbf{ \Xi} \Theta(y_i)\|\|^2$$
$$\prod_i p (\dot{y_i}|\mathbf{\Xi},  y_i)= p(\mathbf{\dot X|\Xi}, \mathbf{X})$$
where we're treating the $\dot y$ calculated in the differentiation step as measurements (and $y$, smoothed in that same step, as some ground truth).  The difference from the first case is that the function library at time $i$ is calculated off of a value measured at time $i$, rather than simulated from $t=0$.  For discrete cases, that same issue leads to a different probability expression:
$$p (y_i|\mathbf{\Xi}, y_{i-1}) = \frac{1}{\dots}\|\|y_i -\mathbf{ \Xi} \Theta(y_{i-1})\|\|^2$$
$$\prod_i p (y_i|\mathbf{\Xi},  y_{i-1})= p(y_n|\mathbf{\Xi}, y_{0})$$

So... sorry if you've already addressed this in the other branch, or if I've made a mistake here.  But I needed to work through it all to understand it.

Jacob-Stevens-Haas commented:
So TL; DR from previous comment is that I don't think any change is needed here on the PR to address discrete vs continuous.  Hirsh only deal with continuous, but I don't think the analogous data likelihood for discrete cases is what we use either.

Jacob-Stevens-Haas commented:
First testing thoughts:
#### Add SBR to tests of multiple optimizers
```
test_supports_multiple_targets
test_fit
test_not_fitted
test_complexity_not_fitted
test_illegal_unbias
test_normalize_columns
test_pickle
```
#### SBR-specific
* Something to test guard code for bad argument values
* Something that directly tests horsehoe prior (`_sample_reg_horseshoe`) in a very simple case
* Something that tests overall SBR in a very simple case

mikkelbue commented:
@Jacob-Stevens-Haas thanks for all you comments and thoughts!

>One of the thorniest issues I've dealt with in pysindy is the multiplicity of classes that involve repeating code to make a particular use case or method work, often signaled by if isinstance(... calls.

I completely agree that this type of design pattern does not spark joy :smiley:. The branch I have linked to is an attempt to make the integral approach work without changing anything in the core library. If I would do some refactoring, I would probably equip the feature libraries with a minimal `_transform` method that would implement the core operation of the transform but without any checks, so that `_transform` could work for both `AxesArray` and `jax.numpy.array`, and then call that method directly from the `diffrax` integrator as well as from the `transform` method of the feature library. But that requires some careful engineering... I am happy to contribute to the discussion of how `pysindy` could be refactored to better support e.g. Bayesian SINDy.

> But anyways, to illustrate my point, here's Eq 3.3, ... But I needed to work through it all to understand it.

You are absolutely correct! I think I got confused because for the discrete case, since here the left-hand-side $x_i$ are members of the data space $x_i \in \mathcal D$, while that is not true for $\dot x_i$. So the inferred $\sigma$ has the correct unit for the discrete case, even when not composing multiple steps. If the steps are not composed (as in the current implementation) the underlying assumption is that the steps are independent, while when composing (or integrating), the assumption is that each step must be consistent with *all* the previous steps. This is a stronger constraint, which is also reflected in the more concentrated posteriors in [the example](https://github.com/mikkelbue/pysindy/blob/bayesian_sindy_diffrax/examples/19_bayesian_sindy/example.ipynb) that showcases both approaches.

> we're treating the $\dot z$ calculated in the differentiation step as measurements (and $z_i$, smoothed in that same step, as measurements replacing the original measurements).

I wasn't aware that the differentiation method can also smooth the states! This is something to consider. I get why that is useful for "frequentist" SINDy, but it would change the character of the observation noise, which would be reflected in  a different posterior distribution. It's not done for the default `FiniteDifference`, is it?

I agree we should go ahead and get this merged. My concern with the non-discrete case (as mentioned before)  is that $\dot x_i \notin \mathcal D$, so the inferred $\sigma$ is not clearly interpretable. But I think that is okay as long as it's clearly explained in the docstring. having this wrapped up will also make the next steps clearer. I will have a look at writing the appropriate tests some time this week.

Jacob-Stevens-Haas commented:
Ah, that's a good point about $\sigma$ and $\mathcal D$.  Just goes to show how deep it's possible to go in a paper.

mikkelbue commented:
@Jacob-Stevens-Haas 

Starting to work on the tests. The GitHub workflow is giving me a strange error though:
```
sphinx.errors.SphinxWarning: /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pysindy/optimizers/sbr.py:docstring of pysindy.optimizers.sbr.SBR:7:Unknown target name: "mcmc".

Warning, treated as error:
/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pysindy/optimizers/sbr.py:docstring of pysindy.optimizers.sbr.SBR:7:Unknown target name: "mcmc".
```

I think it's because it's interpreting the `mcmc_` as a reference but I don't know what to do about it.

mikkelbue commented:
> I think it's because it's interpreting the `mcmc_` as a reference but I don't know what to do about it.

Never mind, I figured it out.

mikkelbue commented:
Test writing progress:

**Add SBR to tests of multiple optimizers**
[] test_supports_multiple_targets
[] test_fit
[] test_not_fitted
[] test_complexity_not_fitted
[] test_illegal_unbias
[] test_normalize_columns
[] test_pickle

**SBR-specific**
[] Something to test guard code for bad argument values
[] Something that directly tests horsehoe prior (_sample_reg_horseshoe) in a very simple case
[] Something that tests overall SBR in a very simple case

mikkelbue commented:
>  test_pickle

This is now added by refactoring the existing `test_pickle` using `@pytest.mark.parametrize`.

> Something that directly tests horsehoe prior (_sample_reg_horseshoe) in a very simple case

This is slightly problematic, since if called outside a model context, the `numpyro.sample` functions would need to take an additional argument `key` to produce a sample. I can adapt the function to optionally take this, but that feels a little roundabout just to get a unit test set up. What do you think?

> Something that tests overall SBR in a very simple case

This is now done as `test_sbr_fit`. It just asserts whether `SBR` has an `mcmc_` attribute after fitting and whether it contains samples of the correct variables.

Jacob-Stevens-Haas commented:
> I can adapt the function to optionally take this, but that feels a little roundabout just to get a unit test set up. What do you think?

No, that makes sense. I always wonder about what the right way to test things things is (and thereby, understand), but having read more about reg horseshe, I agree the roundabout work is not necessary here.  I read a bit more about the reg-horseshoe, so I'll add some more info in the docstrings back in in a suggested edit.

> This is now done as `test_sbr_fit`. It just asserts whether SBR has an mcmc_ attribute after fitting and whether it contains samples of the correct variables.

I meant even simpler setup, but more rigorous test, a la [`test_sparse_subset`][1].  Try this one below:

```python
def test_sbr_fit():
    x = np.eye(2)
    y = np.array([[1], [1e-4]])
    expected = np.array([[1], [0]])
    result = SBR(num_warmup=10, num_samples=10).fit(x, y).coef_
    np.testing.assert_allclose(result, expected, atol=1e-1)
```

This is a good example of what it takes to get SBR to succeed in the simplest of cases, and I'm looking into why it's failing.

[1]: https://github.com/dynamicslab/pysindy/blob/b497134041c0fa5c991a5386213de2db92612bb4/test/test_optimizers.py#L1115

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,189 @@
+    horseshoe prior over the SINDy coefficients to achieve sparsification.
+
+    The horseshoe prior contains a "spike" of nonzero probability at the
+    origin, and a "slab" of distribution in cases where a coefficient is nonzero.
```
```suggestion
    origin, and a Student's-T-shaped "slab" of distribution in cases where a
    coefficient is nonzero.

```

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,189 @@
+        value increases the sparsity of the SINDy coefficients.
+
+    slab_shape_nu : float, optional (default 4)
+        Controls spread of slab.  For values less than 4,
+        the kurtosis of of nonzero coefficients is undefined.  As  the value
+        increases past 4, for higher values, the variance and kurtosis approach
+        :math:`s` and :math:`s^2`, respectively
```
```suggestion
        Along with ``slab_shape_s``, controls tails of nonzero coefficients.
        Specifically, degrees of freedom for Student's-T-shaped slab.
        Higher values decrease excess kurtosis to zero, lower values >= 4
        increase kurtosis to infinity. 
```

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,189 @@
+        :math:`s` and :math:`s^2`, respectively
+
+    slab_shape_s : float, optional (default 2)
+        Controls spread of slab.  Higher values lead to more spread
+        out nonzero coefficients.
```
```suggestion
        Along with ``slab_shape_nu``, controls standard deviation of nonzero
        coefficients.
```

Jacob-Stevens-Haas commented on pysindy/optimizers/sbr.py:
```diff
@@ -0,0 +1,189 @@
+        return mcmc
+
+
+def _sample_reg_horseshoe(tau, c_sq, shape):
```
```suggestion
def _sample_reg_horseshoe(tau, c_sq, shape):
def _sample_reg_horseshoe(tau: float, c_sq: float, shape: tuple[int, ...]):
    """Create a regularized horseshoe distribution

    The regularized horseshoe distribution behaves like a horseshoe prior when
    shrinkage ``lamb`` is small, but behaves like a gaussian slab of variance
    ``c_sq`` when ``lamb`` is big or a Student T-slab when ``c_sq`` is itself
    an inverse Gamma.

    For original work, inluding interpretation of the coefficients, see:

    Piironen, J., and Vehtari, A. (2017). Sparsity Information and
    Regularization in the Horseshoe and Other Shrinkage Priors. Eletronic Journal
    of Statistics Vol. 11 pp 5018-5051. https://doi.org/10.1214/17-EJS1337SI
    """
```

Jacob-Stevens-Haas reviewed (changes requested):
All suggested docstring changes based on looking into parameters to get the test case to fit.

Jacob-Stevens-Haas commented:
Ok, I've got it but I have to run, will post it a bit later.  `x` and `y` in my code need more samples - even dropping the noise prior near zero doesn't do it.

Jacob-Stevens-Haas commented:
For the notebooks, 
1) Combine them (just one notebook file) as example.ipynb
2) Save as a python file
3) `cd examples`, then  `python publish_notebook.py 19_bayesian_sindy`.
4) Run `pytest --durations=10 --external_notebook 19_bayesian_sindy` to see if it tests
(or you can run `pytest --durations=10 -m "notebook" -k test_notebook` and see if it gets picked up)

mikkelbue commented:
@Jacob-Stevens-Haas 
> So that's all for testing. For the notebooks, ...

All done. However ` pytest --durations=10 --external_notebook 19_bayesian_sindy` fails with `pytest: error: unrecognized arguments: --external_notebook` and I can't find any documentation for that argument.

Jacob-Stevens-Haas commented:
LGTM, and thank you for your dedication on this!  I'll look into that pytest issue separately - pytest allows you create command line arguments in conftest.py, which is what I did for `--external-notebook`, but that was made in pytest 5.X and we're now on 7.X (actually, 8.X, but it's incompatible with one of the plugins we use).  It existed before the split to `example.py`/`example.ipynb`, so probably not necessary anymore.
